pydantic/_internal/_decorators.py

```diff
@@ -299,33 +392,36 @@ def __call__(self, __value: Any, values: dict[str, Any], **kwargs: Any) -> Any:
 ]
 
 
-def make_generic_validator(
-    validator: V1Validator | OnlyValueValidator | GeneralValidatorFunction, mode: str
-) -> GeneralValidatorFunction:
-    """
-    In order to support different signatures, including deprecated validator signatures from v1,
-    we introspect the function signature and wrap it in a parent function that has a signature
-    compatible with pydantic_core
-    """
-    sig = signature(validator)
-
-    def _warn_v1_validator() -> None:
-        warnings.warn(
-            'Validator signatures using the `values` keyword argument or `**kwargs` are no longer supported.'
-            ' Please use an `info: pydantic.ValidationInfo` as the second positional argument instead.'
-            ' This compatibility shim may be removed in a future minor release of Pydantic v2.X',
-            DeprecationWarning,
-            # The stacklevel parameter makes the warning show up as coming from the User's code instead
-            # of our internal implementation
-            # Since this just goes up the stack the source location will appear as
-            # class TheModelName(BaseModel):
-            # Which is good enough for now in terms of helping users locate the issue
-            # In the future maybe we can capture the calling module and line number in the
-            # @validator decorator and use that here to be more accurate about where the issue is
-            # But that may not be 100% reliable, so tabling for now.
-            stacklevel=6,
-        )
+V1_VALIDATOR_VALID_SIGNATURES = """\
+def f1(value: Any) -> Any: ...
+def f2(value: Any, values: Dict[str, Any]) -> Any: ...
+
+class Model(BaseModel):
+    x: int
+
+    @validator('x')
+    @classmethod  # optional
+    def val_x1(cls, value: Any) -> Any: ...
+
+    @validator('x')
+    @classmethod  # optional
+    def val_x2(cls, value: Any, values: Dict[str, Any]) -> Any: ...
+
+    @validator('x')
+    @staticmethod  # required
+    def val_x3(value: Any) -> Any: ...
 
+    @validator('x')
+    @staticmethod  # required
+    def val_x4(value: Any, values: Dict[str, Any]) -> Any: ...
+
+    val_x5 = validator('x')(f1)
+    val_x6 = validator('x')(f2)
+"""
+
+
+def make_generic_v1_field_validator(validator: V1Validator) -> FieldValidatorFunction:
```

Comment:
Also, much of this function seems uncovered by tests. I'm guessing it wouldn't be hard to change that by re-adding tests with the old stuff and wrapping them in deprecation-warning-catchers. You might be able to do something like what I did with `tests.test_construction.copy_method` where it parametrizes over both, and ensures a deprecation warning is raised for the deprecated one (you'd only use it where it was runtime-viable to use either of the validator or field_validator decorators)

Reply:
I'll add tests to cover these cases.